File: src/model/rules/RuleImplReversi.java
package model.rules;

import model.rules.displayRule.DisplayRule;
import model.rules.displayRule.DisplayRuleImplReversi;
import model.rules.gameRule.GameRule;
import model.rules.gameRule.GameRuleImplReversi;
import model.rules.inputRule.InputRule;
import model.rules.inputRule.InputRuleImplMonochrome;
import model.rules.inputRule.InputRuleImplReversi;

public class RuleImplReversi implements Rule {

    private static final RuleImplReversi instance = new RuleImplReversi();

    public static RuleImplReversi getRule() {
        return instance;
    }

    private RuleImplReversi() {}

    final InputRule inputRule = InputRuleImplReversi.getInputRule();
    final GameRule gameRule = GameRuleImplReversi.getGameRule();
    final DisplayRule displayRule = DisplayRuleImplReversi.getDisplayRule();

    @Override
    public GameRule getGameRule() {
        return gameRule;
    }

    @Override
    public InputRule getInputRule() {
        return inputRule;
    }

    @Override
    public DisplayRule getDisplayRule() {
        return displayRule;
    }

    public static final String name = "reversi";
    @Override
    public String getName() { return name; }
}

File: src/model/rules/RuleImplReversi.kt
package model.rules

import model.rules.displayRule.DisplayRule
import model.rules.displayRule.DisplayRuleImplReversi
import model.rules.gameRule.GameRule
import model.rules.gameRule.GameRuleImplReversi
import model.rules.inputRule.InputRule
import model.rules.inputRule.InputRuleImplReversi

class RuleImplReversi private constructor() : Rule {
    override val inputRule: InputRule = InputRuleImplReversi.inputRule
    override val gameRule: GameRule = GameRuleImplReversi.gameRule
    override val displayRule: DisplayRule = DisplayRuleImplReversi.displayRule

    override val name: String
        get() = Companion.name

    companion object {
        val rule: RuleImplReversi = RuleImplReversi()

        const val name: String = "reversi"
    }
}
